Applications/Spire/Source/KeyBindings/OrderFieldAdditionalTag.cpp

```diff
@@ -0,0 +1,92 @@
+#include "Spire/KeyBindings/OrderFieldAdditionalTag.hpp"
+#include "Spire/Spire/ArrayListModel.hpp"
+#include "Spire/Ui/CalendarDatePicker.hpp"
+#include "Spire/Ui/DateBox.hpp"
+#include "Spire/Ui/DecimalBox.hpp"
+#include "Spire/Ui/DropDownBox.hpp"
+#include "Spire/Ui/DurationBox.hpp"
+#include "Spire/Ui/IntegerBox.hpp"
+#include "Spire/Ui/MoneyBox.hpp"
+#include "Spire/Ui/QuantityBox.hpp"
+#include "Spire/Ui/TextBox.hpp"
+
+using namespace boost;
+using namespace boost::gregorian;
+using namespace boost::posix_time;
+using namespace Nexus;
+using namespace Spire;
+
+namespace {
+  auto make_quantity_modifiers() {
+    return QHash<Qt::KeyboardModifier, Quantity>(
+      {{Qt::NoModifier, 1}, {Qt::AltModifier, 5}, {Qt::ControlModifier, 10},
+       {Qt::ShiftModifier, 20}});
+  }
+
+  auto make_drop_down_box(const OrderFieldInfoTip::Model& model,
+      const QString& value) {
+    auto list = std::make_shared<ArrayListModel<QString>>();
+    auto current = std::make_shared<LocalValueModel<optional<int>>>();
+    for(auto i = 0; i < std::ssize(model.m_tag.m_arguments); ++i) {
+      auto argument_value =
+        QString::fromStdString(model.m_tag.m_arguments[i].m_value);
+      list->push(argument_value);
+      if(argument_value == value) {
+        current->set(i);
+      }
+    }
+    return new DropDownBox(std::move(list), std::move(current),
+      ListView::default_view_builder);
+  }
+}
+
+OrderFieldAdditionalTag::OrderFieldAdditionalTag(Nexus::Tag tag,
+  OrderFieldInfoTip::Model model)
+  : AdditionalTag(model.m_tag.m_name.c_str(), tag.GetKey()),
+    m_model(std::move(model)),
+    m_value(std::make_shared<LocalValueModel<optional<Nexus::Tag::Type>>>(
+      tag.GetValue())) {}
+
+const OrderFieldInfoTip::Model& OrderFieldAdditionalTag::get_model() const {
+  return m_model;
+}
+
+const std::shared_ptr<AdditionalTagValueModel>&
+    OrderFieldAdditionalTag::make_value() const {
+  return m_value;
```

Comment:
This is not consistent with the API which states:

    /** Makes a new value suitable for this tag. */

Reply:
If the function returns a new value every time it’s called, I think it would be reasonable to return a value instead of a reference, since returning the address of a local object would be invalid.

I propose updating the return type of `AdditionalTag::make_value`:
```
virtual std::shared_ptr<AdditionalTagValueModel>
    make_value() const = 0;
```

Reply:
Good catch, that makes sense.